homeassistant/components/rainmachine.py

```diff
@@ -87,3 +87,46 @@ def setup(hass, config):
     _LOGGER.debug('Setup complete')
 
     return True
+
+
+class RainMachine(object):
+    """Define a generic RainMachine object."""
+
+    def __init__(self, client):
+        """Initialize."""
+        self.client = client
+        self.device_mac = self.client.provision.wifi()['macAddress']
+
+
+class RainMachineEntity(Entity):
+    """Define a generic RainMachine entity."""
+
+    def __init__(self,
+                 rainmachine,
+                 rainmachine_type,
+                 rainmachine_entity_id,
+                 icon=DEFAULT_ICON):
+        """Initialize."""
+        self._attrs = {ATTR_ATTRIBUTION: DEFAULT_ATTRIBUTION}
+        self._icon = icon
+        self._rainmachine_type = rainmachine_type
+        self._rainmachine_entity_id = rainmachine_entity_id
+        self.rainmachine = rainmachine
+
+    @property
+    def device_state_attributes(self) -> dict:
+        """Return the state attributes."""
+        return self._attrs
+
+    @property
+    def icon(self) -> str:
+        """Return the icon."""
+        return self._icon
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique, HASS-friendly identifier for this entity."""
+        return '{0}_{1}_{2}'.format(
+            self.rainmachine.device_mac.replace(
```

Comment:
We should just use mac and type. I don't know what entity id is ?

Reply:
MAC and type aren't sufficient – in that case, every program would have the same unique ID (e.g., "abcdef1234_program"). Same with zones. `entity_id` is the ID of that specific program or zone, which gives results like "abcdef1234_program_2". Let me know if I'm thinking about this wrong or doing the job in a confusing manner.

Reply:
That makes sense. Got confused because Home Assistant also has the concept of entity ids.